Add allowedGroups to UserConfig in IdZ configuration

klaus-sap opened the pull request:
https://github.com/cloudfoundry/uaa/issues/2505 -> Proposal 3.
We want to add an optional attribute "allowedGroups" to the UserConfig in IdZ configuration. If this attribute is not set (i.e., allowedGroups == null), then all groups are allowed as before. If this attribute contains an array of groups, then only the groups from defaultGroups plus allowedGroups are allowed. All other groups can neither be created nor assigned to users.

klaus-sap committed:
fix sonar issue: Utility classes should not have public constructors

klaus-sap committed:
Merge branch 'cloudfoundry:develop' into develop

klaus-sap committed:
fix: possibly unnecessary method call

klaus-sap committed:
fix: define Java version

klaus-sap committed:
handle in separate PR

klaus-sap committed:
Merge branch 'cloudfoundry:develop' into develop

klaus-sap committed:
Merge branch 'cloudfoundry:develop' into develop

klaus-sap committed:
feature: add attribute userConfig.allowedGroups to IdZ

klaus-sap committed:
Merge branch 'cloudfoundry:develop' into develop

strehle commented:
@klaus-sap here you can see the smells from sonar
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2606

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -222,8 +227,29 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
+        }
+        IdentityZone currentZone = IdentityZoneHolder.get();
+        try {
+            zoneAllowedGroups = (zoneId.equals(currentZone.getId())) ?
```
You should add this only if currentZone,.getId != "uaa" , because the default zone we dont want /need this check here?

klaus-sap replied:
I would prefer to handle the configuration of all zones in the same way

strehle commented on model/src/main/java/org/cloudfoundry/identity/uaa/zone/UserConfig.java:
```diff
@@ -27,4 +27,14 @@ public List<String> getDefaultGroups() {
     }
+
+    private List<String> allowedGroups = null;
+
```
Combine the default and the allowed Groups
e.g. if you have entries in defaultGroups, they are automatically in allowed as far as I understood?

klaus-sap replied:
done

strehle reviewed (commented):
This should work in all zones != uaa

strehle committed:
revert format changes

strehle committed:
revert format changes

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -74,6 +78,7 @@ public JdbcScimGroupProvisioning(
         super(jdbcTemplate, pagingListFactory, new ScimGroupRowMapper());
 
         this.jdbcTemplate = jdbcTemplate;
+        jdbcIdentityZoneProvisioning = new JdbcIdentityZoneProvisioning(jdbcTemplate); // TODO inject
```
please enhance
https://github.com/cloudfoundry/uaa/blob/develop/uaa/src/main/webapp/WEB-INF/spring/scim-endpoints.xml#L112
with jdbcIdentityZoneProvisioning

klaus-sap replied:
no bean jdbcIdentityZoneProvisioning found for ref

klaus-sap committed:
fix Sonar warnings

klaus-sap committed:
remove TODO

strehle committed:
create new needed bean similar than the other beans created

strehle committed:
is an optional entry in identity zone configuration

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into klaus-develop

strehle committed:
fix sonar smells

strehle reviewed (dismissed):
Open
a) combine default and allowed groups. So from understanding, if you have default groups defined, then these entries are automatically part of allowedGroups. So this behaviour should be implemented.
b) Check if the limitation should be really done always or only in Identity Zone NOT EQUAL to 'uaa'

klaus-sap committed:
combine default and allowed groups

klaus-sap commented:
> Open a) combine default and allowed groups. So from understanding, if you have default groups defined, then these entries are automatically part of allowedGroups. So this behaviour should be implemented. b) Check if the limitation should be really done always or only in Identity Zone NOT EQUAL to 'uaa'

a) commit [6e6ab18](https://github.com/cloudfoundry/uaa/pull/2606/commits/6e6ab189b2be5f60f2b3200ede998a04c5253aa8)
b) I would prefer to handle the IdZ configuration of all zones in the same way. If you don't want allowedGroups in the "uaa" zone, simply don't configure them

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaAuthorizationRequestManager.java:
```diff
@@ -249,7 +249,11 @@ private Set<String> checkUserScopes(Set<String> requestedScopes,
+        Collection<String> allowedScopes = IdentityZoneHolder.get().getConfig().getUserConfig().resultingAllowedGroups();
         allowed.addAll(defaultScopes);
+        if (allowedScopes != null) {
+            allowed.retainAll(allowedScopes);
```
Test missing? or did not found one
Situation. Your authorities contain 3 entries but in your zone you allow only 1 of them

klaus-sap replied:
added testScopesIncludesAllowedAuthoritiesForUser() [a464bab](https://github.com/cloudfoundry/uaa/pull/2606/commits/a464babd7399d39fa3eaa6778b9b12dec00ecee7)

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -222,10 +231,32 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
+    private Set<String> getAllowedUserGroups(String zoneId) {
+        Set<String> zoneAllowedGroups = null; // default: all groups allowed
+        if (!hasText(zoneId)) {
+            return zoneAllowedGroups;
```
No Test found for this

klaus-sap replied:
zoneId must not be null, changed logic: [370be52](https://github.com/cloudfoundry/uaa/pull/2606/commits/370be522393608696adab86ba3888269b8ed235d)

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -222,10 +231,32 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
+        }
         final String id = UUID.randomUUID().toString();
-        logger.debug("creating new group with id: " + id);
+        logger.debug("creating new group with id: %s", id);
```
here you can do {} instead of %s

klaus-sap replied:
done

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -248,6 +279,11 @@ public ScimGroup create(final ScimGroup group, final String zoneId) throws Inval
             ScimResourceNotFoundException {
+        Set<String> allowedGroups = getAllowedUserGroups(zoneId);
+        if ((allowedGroups != null) && (!allowedGroups.contains(group.getDisplayName()))) {
+            throw new InvalidScimResourceException("The group with displayName: " + group.getDisplayName()
```
No test found with this error

klaus-sap replied:
check error message in cannotUpdateNotAllowedGroup: [370be52](https://github.com/cloudfoundry/uaa/pull/2606/commits/370be522393608696adab86ba3888269b8ed235d)

strehle commented on model/src/main/java/org/cloudfoundry/identity/uaa/zone/UserConfig.java:
```diff
@@ -27,4 +29,26 @@ public List<String> getDefaultGroups() {
+
+    // return defaultGroups plus allowedGroups
+    public Set<String> resultingAllowedGroups() {
+        if (allowedGroups == null) {
```
Test coverage in this area should be increased

klaus-sap replied:
added unit test UserConfigTest.java with commit [8cde93e](https://github.com/cloudfoundry/uaa/pull/2606/commits/8cde93e0abb522f5a23552c547eb9d7228715a85)

strehle reviewed (commented):
see here new sonar

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2606

coverage
https://sonarcloud.io/component_measures?id=cloudfoundry-identity-parent&pullRequest=2606&metric=new_coverage&view=list

smells
not all need to be fixed, but please check them
https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&sinceLeakPeriod=true&pullRequest=2606&id=cloudfoundry-identity-parent

strehle commented:
@klaus-sap Please add some scenario tests into
https://github.com/cloudfoundry/uaa/blob/develop/uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ScimGroupEndpointsIntegrationTests.java

you can create an extra zone or stay in default , btw. zone creation https://github.com/cloudfoundry/uaa/blob/develop/uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/util/IntegrationTestUtils.java#L671

Scenario I see
a) Add allowedGroups to zone and perform positive and negative test , means create the group 
b) Update or deleted one entry in defaultGroups (because this has influence to allowedGroups) and then do the same create the groups and verify result

klaus-sap committed:
fix Sonar warnings

klaus-sap committed:
add test class

klaus-sap committed:
add testResultingAllowedGroups

klaus-sap committed:
add testScopesIncludesAllowedAuthoritiesForUser

klaus-sap committed:
zoneId must not be null

klaus-sap committed:
use right asserts

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -222,10 +231,32 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
+    @SuppressWarnings("java:S1874")
+    private Set<String> getAllowedUserGroups(String zoneId) {
+        if (zoneId == null) {
+            throw new IllegalArgumentException("zoneId is null");
```
this here is not tested, but if you change 
 (zoneId.equals(currentZone.getId())) 
to
 (currentZone.getId(equals(zoneId))

you dont need the null check for zoneId

klaus-sap replied:
done with [ab2d5cf](https://github.com/cloudfoundry/uaa/pull/2606/commits/ab2d5cf74037ce00effc502395c8b2578be87c2e)
(line 241) "jdbcIdentityZoneProvisioning.retrieve(zoneId)" seems to be null-safe as well

klaus-sap committed:
remove null-check

klaus-sap committed:
add integration tests

klaus-sap committed:
fix createNotAllowedGroupFails

klaus-sap committed:
introduce list allowedGroups

klaus-sap committed:
update allowedGroups on server

klaus-sap committed:
pass IdZ id

klaus-sap committed:
determine zone id

strehle committed:
update integration tests

We have 4 tests
2 positive and 2 negative
2 using allowedGroups
2 relying on defaultGroups because allowedGroups is empty

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into integrationTest

strehle committed:
cleanup in integration tests, fixes mftop

refactor to omit dublicates

strehle reviewed (approved):
### sonar ok,
- almost 100 coverage.
- Integration Tests available

https://sonarcloud.io/component_measures?id=cloudfoundry-identity-parent&pullRequest=2606&metric=new_coverage&view=list

@hsinn0 @Tallicia FYI please check who can do a review from VMware side in next week , thanks

strehle committed:
Merge remote-tracking branch 'origin/develop' into klaus-rebase

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/UserConfigValidator.java:
```diff
@@ -0,0 +1,23 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UserConfigValidator {
+    private static Logger logger = LoggerFactory.getLogger(UserConfigValidator.class);
+
+    // add a private constructor to hide the implicit public one
+    private UserConfigValidator() {
+    }
+
+    public static void validate(UserConfig config) throws InvalidIdentityZoneConfigurationException {
+        Set<String> allowedGroups = (config == null) ? null : config.resultingAllowedGroups();
+        if ((allowedGroups != null) && (allowedGroups.isEmpty())) {
+            String message = "At least one group must be allowed";
+            logger.error(message);
+            throw new InvalidIdentityZoneConfigurationException(message);
+        }
+    }
+}
```
It feels like this class doesn't need to exist, and the method could be part of `UserConfig`. I like to avoid util classes whenever I can.

strehle replied:
we will have more to validate later, so this validator classes are a pattern in other parts of UAA

bruce-ricard replied:
> validator classes are a pattern in other parts of UAA

I know they are a commonly used pattern in UAA. I think that this is a bad pattern, I'm trying to move away from it.

IMHO it would be much better for this static `validate()` method to be moved as an instance method of the `UserConfig` class.

But I can understand if you really want to have the validation code in another class. If that's the case, let me suggest this pattern:
In UserConfig, you have:

```
public static void validate(UserConfigValidator validator) {
    validator.validate(this)
}
```
And in the `UserConfigValidator`, the method becomes an instance one. The class is no longer a "util" class with static methods. 

I think that this is a common OO design pattern, I just don't know its name.

klaus-sap replied:
I would prefer to stay with a separate class UserConfig, because we now have a matching test class UserConfigTest. There are no member variables in UserConfig, so I would like to leave it static

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/UserConfigValidator.java:
```diff
@@ -0,0 +1,23 @@
+        Set<String> allowedGroups = (config == null) ? null : config.resultingAllowedGroups();
+        if ((allowedGroups != null) && (allowedGroups.isEmpty())) {
+            String message = "At least one group must be allowed";
+            logger.error(message);
```
This is not an error. This is info. Error means the UAA is having an error. Here it's behaving as expected.

klaus-sap replied:
changed log level with cc27f6661a541d1462b07af49c245803869d4bac

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/UserConfigValidator.java:
```diff
@@ -0,0 +1,23 @@
+        if ((allowedGroups != null) && (allowedGroups.isEmpty())) {
+            String message = "At least one group must be allowed";
+            logger.error(message);
+            throw new InvalidIdentityZoneConfigurationException(message);
```
I'm surprised that Sonar isn't flagging this. Usually it says that you should either log the error, or throw the exception.

Now that I think about it, I think that Sonar only catches it when it's about catching and rethrowing an exception.

But I still think that it applies here. If feels like your message is going to be logged 2 times. Once here, and wherever the exception caught.

klaus-sap replied:
changed with ef94dea12c7af9305d06317acb0daf34fd5c7c6f

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -324,4 +349,12 @@ protected void validateOrderBy(String orderBy) throws IllegalArgumentException {
         super.validateOrderBy(orderBy, GROUP_FIELDS);
     }
 
+    private void validateAllowedUserGroups(String zoneId, ScimGroup group) {
```
You are only validating the allowed groups in `create` and `update`. It seems to me like it would be possible for a group to have been created in the past with certain user groups, then the admin sets the userconfig to specify certain allowed user groups, and this object still contains the old user groups that haven't been validated. And will never be validated until the object is updated, which could be never.

strehle replied:
you want the validation in token creation ?

bruce-ricard replied:
I am not sure if that makes sense. That would require the validation to be done at every single token creation time, that feels wrong from the performance point of view. But it seems that you are correct, that it would solve this issue.

I was thinking of another idea: whenever the user config is updated, we need to revalidate all groups. That would make the config update more computationally heavy, but that is a lot less frequent than token creation.

klaus-sap replied:
Ok, this would be a change in "Update Identity Zone", right? We might add a check if there are groups which would be not allowed anymore after updating the IdZ. In this case, we would raise an UnprocessableEntityException

klaus-sap replied:
Added check: [56adc44](https://github.com/cloudfoundry/uaa/pull/2606/commits/56adc44dbf4b0281bf8044e852fce5ce3d60253a)

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -220,9 +229,24 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
         }
     }
 
+    @SuppressWarnings("java:S1874")
```
I see lots of these in this PR. That feels like a code smell. Why are we suppressing all these warning?

Also, I don't know what warning this is suppressing. I tried googling this number, I can't find anything. It's not great for being able to understand the code.

strehle replied:
Ok, but this is a old discussion old. Must of the S1874 warnings from from deprecated code which we cause, e.g. IdentityZoneHoler or https://github.com/cloudfoundry/uaa/milestone/38 but this annotation simply cleanup  
our own deprecations.....  why is IdentityZoneHoler deprecated but nobody cares about it or removes it

bruce-ricard replied:
IMO, we should suppress the warning when we believe the warning is incorrect. We shouldn't suppress the warnings because we don't like the noise, and because we don't want to fix the underlying issue. 

Here, a real issue is reported by Sonar. We shouldn't hide it.

I am not saying that we need to fix the deprecation problem right now. But IMO we need to remove the tags, and allow Sonar to show that there is a problem.

strehle replied:
removed

bruce-ricard commented on model/src/main/java/org/cloudfoundry/identity/uaa/zone/UserConfig.java:
```diff
@@ -27,4 +29,27 @@ public List<String> getDefaultGroups() {
+    @SuppressWarnings("java:S1168")
+    public Set<String> resultingAllowedGroups() {
+        if (allowedGroups == null) {
+            return null; // null = all groups allowed
```
I like to try to avoid `null`s whenever I can. I think that here we can. At the very least, it's quite confusing to the reader that `null` means when they are calling this method and it returns `null`, assuming that they remember to do the null check.

I have 2 different suggestions:
1. Use `Optional`. The type of this field would be `Optional<Set<String>>`. It solves the null problem, but it leaves open the interpretation of what `empty` is.
2. Have another method `public bool areAllUserGroupsAllowed` to catch that case. If that returns false, you can call the `resultingAllowedGroups` to see which groups are allowed. I understand that it is a bit heavy, but it has the advantage of being very clear and readable.

strehle replied:
I vote for 2, which might simplify it

klaus-sap replied:
ok, but what should resultingAllowedGroups() return if areAllUserGroupsAllowed == true? An empty optional?

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -220,9 +229,24 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
+        Set<String> zoneAllowedGroups = null; // default: all groups allowed
+        try {
+            IdentityZone currentZone = IdentityZoneHolder.get();
+            zoneAllowedGroups = (currentZone.getId().equals(zoneId)) ?
+                currentZone.getConfig().getUserConfig().resultingAllowedGroups() :
+                jdbcIdentityZoneProvisioning.retrieve(zoneId).getConfig().getUserConfig().resultingAllowedGroups();
```
Why do you have the conditional here? Wouldn't it be simpler to just always do the `retrieve(zoneId)`, instead of checking if we are in the current zone?

strehle replied:
Hi @bruce-ricard ok, so vote to stay here with the logic . We may can reactor readability, but the current zone is in the local thread already, no DB lookup needed, but other zones need to be retrieved from DB. if we simply retrieve from DB, then we have always a DB lookup

bruce-ricard replied:
OK so it's a performance optimization. It's what I was thinking.

I like to try to make code self explanatory, whenever possible. For example here, I don't want future readers of the code to ask themselves this question I asked, until the end of times. 
Typically, I would have created a new method for this operation, and tried to pass the performance optimization explanation to the reader through good variable and method names.

Another different argument would be to say: this method is only called at group update and creation time. This is quite rare. I don't think that we should care about a single DB lookup made at group write time.

Something else: we do the test `currentZone.getId().equals(zoneId)`, and on the next line we get the current zone config. Isn't it possible that between these 2 instructions, another thread changes the current zone, and we end up looking at allowed groups from another zone?

klaus-sap replied:
Regarding the last paragraph: the same may happen if you use an if-statement instead of a conditional. And the time elapsed between the two instructions is really tiny :-)

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java:
```diff
@@ -324,4 +349,12 @@ protected void validateOrderBy(String orderBy) throws IllegalArgumentException {
 
+    private void validateAllowedUserGroups(String zoneId, ScimGroup group) {
+        Set<String> allowedGroups = getAllowedUserGroups(zoneId);
+        if ((allowedGroups != null) && (!allowedGroups.contains(group.getDisplayName()))) {
```
I don't like that we are doing the filtering based on the group display name. That feels quite dangerous.

Is it because in this specific case the "display name" is actually the unique string that represents it as well? If that's the case, could we then rename that method to be `group.identifier()` instead of `getDisplayName()`?

strehle replied:
ok, but the groups in UAA what in reality is used. UAA uses the SCIM group for authorizations and the display name is the scope name. The identitfier is a not used entity , also in other usage of it

bruce-ricard reviewed (dismissed):
I have one major concern about the validation only happening at group write time, see inline comments.

And a couple sub-major implementation design comments.

klaus-sap committed:
change log level to info

klaus-sap committed:
remove log message

klaus-sap committed:
check for groups which would be not allowed after the update

klaus-sap committed:
must not remove default group from configuration

strehle committed:
remove sonar comment

hsinn0 commented:
@bruce-ricard & I agreed that I would take over this PR review from our side as he will be unavailable for a couple of weeks. I am approving this PR.
